Add question type that shows preview and prints to Android-connected printer

grzesiek2010 opened the pull request:
Closes #5831

#### Why is this the best possible solution? Were any other approaches considered?
We have decided to use the [Android printing API](https://developer.android.com/training/printing) because it's versatile and should work with different printers. The API offers three different ways of printing and the one we have chosen is with HTML because in our case we can relatively easily build such HTML docs using the `calculation` column and the [concat function](https://docs.getodk.org/form-operators-functions/#concat). The sample from the form attached below looks like:
```
concat("
<qrcode width=’150’ height=’150’>First name: ", ${first_name}, "Last name: ", ${last_name}, "Age: ", ${age}, "Email: ", ${email}, "Phone number: ", ${phone_number}, "</qrcode>
</br>
<img width=’150’ height=’150’ src=’”, ${photo}, “’>
<table>
   <tr>
      <td>First name</td>
      <td>", ${first_name}, "</td>
   </tr>
   <tr>
      <td>Last name</td>
      <td>", ${last_name}, "</td>
   </tr>
   <tr>
      <td>Age</td>
      <td>", ${age}, "</td>
   </tr>
   <tr>
      <td>Email</td>
      <td>", ${email}, "</td>
   </tr>
   <tr>
      <td>Phone number</td>
      <td>", ${phone_number}, "</td>
   </tr>
</table>
"
)
```

I said it's relatively easy because the more complex the expected output is the more args the function will have. The problem with this function is that every time we want to add an expression that should be evaluated like `${first_name}`, `${last_name}` etc. above we need to add them in as separate args because that's how this function works what makes creating bigger documents difficult and error-prone. 
It would be good to have a new function like:
```
fill_template(“<qrcode>length: ${length} height: ${height}</qrcode>”)
```
That would allow us to have one string, like in kotlin for example. However, this is something we can add separately later.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This introduces a new question type, so the changes seem to be isolated and should not affect other functionalities. Given that, we can (almost) fully focus on testing what's new. The only thing that I would verify to avoid regression is the old printer question type. The old one should be created if the appearance starts with `printer` and the new one if it is just `printer`:
- `printer` then start the new printer displayed
- `printer:printer:org.opendatakit.sensors.ZebraPrinter` the old printer widget should be displayed

#### Do we need any specific form for testing your changes? If so, please attach one.
I've created one sample form: 
[Printing.xlsx](https://github.com/getodk/collect/files/13488794/Printing.xlsx)

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
Yes https://github.com/getodk/docs/issues/1720

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] added a comment above any new strings describing it for translators
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

grzesiek2010 committed:
Added a new simple printer widget

grzesiek2010 committed:
Adde a new module for printing

grzesiek2010 committed:
Added a method for generating QR codes without encoding

grzesiek2010 committed:
Implemented Printer

grzesiek2010 committed:
Implemented parsing html

grzesiek2010 committed:
Trigger printing in PrinterWidget

grzesiek2010 committed:
Added tests for PrinterWidget

grzesiek2010 committed:
Added tests for PrintableHtmlParser

grzesiek2010 committed:
Naming improvements

grzesiek2010 committed:
Added support for printing images

grzesiek2010 committed:
Do not display any value in the hierarchy view for the printer widget

grzesiek2010 committed:
Do not trigger printing if the calculated value is null

grzesiek2010 committed:
Added analytics for the old printer widget

grzesiek2010 committed:
Do not display the option for removing answers in the new printer widget

grzesiek2010 committed:
Added tests

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/widgets/utilities/PrintableHtmlParser.kt:
```diff
@@ -0,0 +1,57 @@
+                val bitmap = BitmapFactory.decodeFile(file.absolutePath)
+                if (bitmap != null) {
+                    val imageData = bitmapToBase64(bitmap)
+                    imgElement.attr("src", "data:image/png;base64,$imageData")
```
Do you have to do this? Is there no way to set the working directory to the media dir when the HTML gets rendered? I don't think it's a big deal either way but would be nice not to have to process big images this way if possible.

grzesiek2010 replied:
Right it is possible. Fixed.

lognaturel commented:
It's magical, @grzesiek2010! 🎉 

I notice it's taking a very long time to load the preview screen after tapping the "Initiate printing" button (I think it should just be "Print"). I also keep getting "ODK Collect isn't responding" after tapping the button including when the preview screen is up and after I've exited out of printing.

I think it's related to the image manipulation. It seems less bad when I don't take a picture (but I do still get not responding bottom sheets every few seconds after I've printed, even when on the main menu screen).

grzesiek2010 committed:
Use absolute paths to include images that alredy exist

grzesiek2010 committed:
Removed redundant method

grzesiek2010 committed:
Display a progress dialog when printing is being prepared

grzesiek2010 committed:
Updated button text

grzesiek2010 commented:
> I notice it's taking a very long time to load the preview screen after tapping the "Initiate printing" button

I've added a progress dialog so that it's clear that something is happening (the HTML is being parsed).

> I think it should just be "Print"

Ok, done.

> I think it's related to the image manipulation. It seems less bad when I don't take a picture (but I do still get not responding bottom sheets every few seconds after I've printed, even when on the main menu screen).

I've updated the way images are added to the HTML. Now I don't convert them to bitmaps and include data URLs but passing a path so maybe it will be faster.

@lognaturel please test if now it works faster and does not cause any ANRs.

grzesiek2010 committed:
Move the logic to viewmodel

lognaturel commented:
Can't reproduce any of the issues I'd highlighted. 👍

seadowg reviewed (changes requested):
This is so cool! I've only had a chance to run through the general structure, and that's mostly looking good to me! A couple of things though:

1. I think it'd be nicer to have `PrinterWidgetViewModel` passed into the widget via the constructor (probably as an interface defined in `widgets` that the ViewModel implements) so that the Activity is in control of constructing it and making observations on its data. This limits the amount of dependencies going into the purely view layer and also means the widget doesn't need to understand anything about the lifecycle it's part of.
2. I'd be interested in if it's possible to do the HTML parsing/manipulation without adding a new dependency (and not increasing the APK size as much) using some of the XML parsing libraries we already use (like `KXML2` for instance).

grzesiek2010 commented:
> I think it'd be nicer to have PrinterWidgetViewModel passed into the widget via the constructor (probably as an interface defined in widgets that the ViewModel implements) so that the Activity is in control of constructing it and making observations on its data. This limits the amount of dependencies going into the purely view layer and also means the widget doesn't need to understand anything about the lifecycle it's part of.

Do you mean building the ViewModel in `FormFillingActivity` then passing to `ODKView` and then from `ODKView` to `WidgetFactory` and from `WidgetFactory` to the widget? Seems like a long path.

> I'd be interested in if it's possible to do the HTML parsing/manipulation without adding a new dependency (and not increasing the APK size as much) using some of the XML parsing libraries we already use (like KXML2 for instance).

I tried kxml2 but it is primarily designed for parsing XML and it didn't work well with HTML. I think we should use a reliable library like Jsoup to make sure even complex HTML constructions work well. We don't want to test such cases manually.

seadowg commented:
> Do you mean building the ViewModel in FormFillingActivity then passing to ODKView and then from ODKView to WidgetFactory and from WidgetFactory to the widget? Seems like a long path.

Yes. We already use this approach for a few ViewModels - we pass the `FormEntryViewModel` and the `FormSaveViewModel` (as a `QuestionMediaManager` which I prefer to the way `FormEntryViewModel` is passed). It's probably yet another sign that `ODKView` should be Fragment (it even has its own lifecycle), but that's probably too big a conversion to do here.

seadowg reviewed (changes requested):
https://github.com/getodk/collect/pull/5854#issuecomment-1875699278

grzesiek2010 committed:
Pass viewmodel to PrinterWidget via constructor

grzesiek2010 committed:
Bumped the max apk size limit

grzesiek2010 commented on .circleci/config.yml:
```diff
@@ -230,8 +230,8 @@ jobs:
 
       - run:
-          name: Check APK size isn't larger than 11.1MB (11639193 * 1024 * 1024)
-          command: if [ $(ls -l collect_app/build/outputs/apk/selfSignedRelease/*.apk | awk '{print $5}') -gt 11639193 ]; then exit 1; fi
```
It looks like 11.1MB should be represented as 11100000 Bytes not 11639193 because I had to bump the limit to 11.8 and 11800000 is enough I don't need 12373196. So the size is in decimal not binary. I was thinking about it reviewing your pr where you have added that change but didn't have time to check it back then.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/widgets/PrinterWidget.kt:
```diff
@@ -0,0 +1,134 @@
+    }
+}
+
+class PrinterWidgetViewModel : ViewModel() {
```
How about moving this elsewhere (`formentry`?) and having an interface in `widgets.interfaces` that this implements.

grzesiek2010 replied:
Done.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/widgets/PrinterWidget.kt:
```diff
@@ -0,0 +1,134 @@
+                document.html()
+            },
+            foreground = { content ->
+                htmlPrinter.print(context, content)
```
Could this happen in the background as well?

grzesiek2010 replied:
No, it would throw an exception:
`java.lang.RuntimeException: WebView cannot be initialized on a thread that has no Looper.`

seadowg commented on collect_app/src/main/java/org/odk/collect/android/widgets/PrinterWidget.kt:
```diff
@@ -0,0 +1,134 @@
+    val isLoading: NonNullLiveData<Boolean> = _isLoading
+
+    fun parseAndPrint(
+        scheduler: Scheduler,
```
I think `Scheduler`, `QuestionMediaManager`, `QRCodeCreator` and `HtmlPrinter` could be constructor args for this class rather than method params.

grzesiek2010 replied:
Could be but maybe not `QuestionMediaManager` since it's a ViewModel itself so passing one ViewModel as a parameter to another one would be weird. Done.

seadowg replied:
Yeah I do agree that that feels a little icky. It's not actually a problem given that they both have the same "lifecycle", but still. This is probably a hint that the `QuestionMediaManager` shouldn't be implemented by a `ViewModel` and should instead be a component shared between them, but that's most likely a change for another time.

seadowg commented on strings/src/main/res/values/strings.xml:
```diff
@@ -216,7 +216,7 @@
     <string name="no_app">The requested application is missing. Please manually enter the reading.</string>
     <string name="null_intent_value">The external application did not provide expected information.</string>
-
+    <string name="print">Print</string>
```
This needs a comment for translators.

grzesiek2010 replied:
Fixed.

seadowg commented on printer/src/main/java/org/odk/collect/printer/HtmlPrinter.kt:
```diff
@@ -0,0 +1,41 @@
+import android.webkit.WebView
+import android.webkit.WebViewClient
+
+class HtmlPrinter {
```
I was in two minds about whether this should be tested or not, but it feels to me like anything with `WebView` is going to be hard to write tests around we'd feel confident about and I feel like you've found a good "edge" here to contain the untested stuff.

grzesiek2010 committed:
Added a comment

grzesiek2010 committed:
Pass scheduler, qrCodeCreator and htmlPrinter via constructor

grzesiek2010 committed:
Moved PrinterWidgetViewModel to formentry package

grzesiek2010 committed:
Removed redundant padding

grzesiek2010 committed:
Return calculated value as answer

grzesiek2010 committed:
Removed unused import

seadowg commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -179,6 +181,12 @@ public class FormHierarchyActivity extends LocalizedActivity implements DeleteRe
     public InstancesRepositoryProvider instancesRepositoryProvider;
 
+    @Inject
+    public QRCodeCreator qrCodeCreator;
```
I don't think either the `QRCodeCreator` or the `HtmlPrinter` need to be injected here as we don't switched them out in tests (unless I'm missing something). They could just be created inline. Same goes for the `FormFillingActivity`.

grzesiek2010 replied:
Fixed.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java:
```diff
@@ -52,6 +52,10 @@ public static String getAnswerText(FormEntryPrompt fep, Context context, FormCon
         IAnswerData data = fep.getAnswerValue();
         final String appearance = fep.getQuestion().getAppearanceAttr();
 
+        if (appearance != null && appearance.equals(Appearances.PRINTER)) {
```
There's no test for this change? Maybe we should rename this method to something like `getHierarchySummary` and write a unit test for it that makes sure the printer questions always return `""`. It'd be nice for the helper to live in the new hierarchy module, but I realise that was actually created in a different PR!

grzesiek2010 replied:
Let's merge https://github.com/getodk/collect/pull/5885 first then and I can add a new test in the same package.

seadowg replied:
I think it would be nice to merge this first so we can get it out as a beta ASAP. Are you able to add a test in #5885?

seadowg commented on collect_app/src/main/java/org/odk/collect/android/widgets/interfaces/Printer.kt:
```diff
@@ -0,0 +1,15 @@
+        context: Context
+    )
+
+    fun isLoading(): NonNullLiveData<Boolean>
```
Do we actually need `NonNullLiveData` any longer? Maybe we should deprecate it?

seadowg commented on collect_app/src/main/java/org/odk/collect/android/formentry/PrinterWidgetViewModel.kt:
```diff
@@ -0,0 +1,80 @@
+    private val qrCodeCreator: QRCodeCreator,
+    private val htmlPrinter: HtmlPrinter
+) : ViewModel(), Printer {
+    private val _isLoading = MutableNonNullLiveData(false)
```
You could use `TrackableWorker` in here now.

grzesiek2010 replied:
Done.

seadowg commented on collect_app/src/test/java/org/odk/collect/android/widgets/PrinterWidgetTest.kt:
```diff
@@ -0,0 +1,76 @@
+import org.odk.collect.qrcode.QRCodeCreator
+import org.odk.collect.testshared.FakeScheduler
+
+class PrinterWidgetTest : QuestionWidgetTest<PrinterWidget, IAnswerData>() {
```
I'd really like to move away from `QuestionWidgetTest` and use shared helpers rather than extension. One example of why would be the no-op tests we end up with here. Do you think it'd be ok to write a version of `PrinterWidgetTest` that doesn't extend `QuestionWidgetTest` or would you rather we just deprecate `QuestionWidgetTest` in follow-up PR?

grzesiek2010 replied:
Hmm I'm not sure... if we introduced a concept of non-question form elements like this new widget (but in a better way without extending not only `QuestionWidgetTest` in tests but generally `QuestionWidget` at all) or the URL widget etc then it would make some sense to rework it but now ok in this case it doesn't look good maybe but forcing writing some tests when we add anew widget generally is a useful thing. Please file a separate issue for that.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/widgets/PrinterWidget.kt:
```diff
@@ -0,0 +1,62 @@
+    init {
+        render()
+
+        MaterialProgressDialogFragment.showOn(
```
This observation should move up to `FormFillingActivity`. Currently, if we had multiple `PrinterWidget` instances on screen, we'd end up making multiple observations that show multiple dialogs (as we call `showOn` for every view creation).

grzesiek2010 replied:
but the dialog will be shown only when `isLoading` is true and it's true when you click the print button. You are not able to click multiple buttons at once so there will never be multiple dialogs displayed right?

seadowg replied:
That's fair, but I still think it makes more sense for the observation to live at the Activity level that than in view creation.

grzesiek2010 replied:
Ok, done.

grzesiek2010 committed:
Do not inject QRCodeCreator and HtmlPrinter

grzesiek2010 committed:
Use the new trackableWorker

grzesiek2010 committed:
Moved observation of loading the document for printing to activity

dbemke commented:
I'm not sure what is the expected when it comes to how the data from questions is represented in the printed version. Here's an example of an answer with new lines and a photo (that changes shape).
![printing](https://github.com/getodk/collect/assets/56479916/fb116e85-6863-45b7-b302-fdcae6ddb728)

What information does the generated QR contain?

grzesiek2010 commented:
> I'm not sure what is the expected when it comes to how the data from questions is represented in the printed version. Here's an example of an answer with new lines and a photo (that changes shape).

Everything is up to you. If you used the form I have attached the code that generates the document contains:
`<img width=’150’ height=’150’ src=’”, ${photo}, “’>`
That means the size of the photo should be 150x150. You can change it.

> What information does the generated QR contain?

again if you used my form it's this part:
`<qrcode width=’150’ height=’150’>First name: ", ${first_name}, "Last name: ", ${last_name}, "Age: ", ${age}, "Email: ", ${email}, "Phone number: ", ${phone_number}, "</qrcode>`

you can add/remove something you can change the size of the code as well.

srujner commented:
I Noticed two strange behaviors:

1) When I set Background Processing to NO, then after generating PDF file and saving it, the PDF file has 0B and I'm not able to open it in any app.

2) When I change the color of the generated PDF file to Black and White and save it, then when opened with Printer+ app the photo is black and white, but when I use the default google app the photo is in color. Is this correct?

grzesiek2010 commented:
Both those issues seem like something not directly related to Collect. If the preview is generated correctly that's where our job ends. Could you try to do the same with other apps that print documents and compare the behavior?

srujner commented:
@grzesiek2010 Checked on different apps and it works in similar way.

**Tested with Success!**

Verified on device with Android 12,13

Verified cases:
- Acceptance Criteria from #5831 has been met;
- Old printer question type;
- Generating pdf file with different widgets;
- Light and dark mode;
- Minimize the app, rotate the screen.

dbemke commented:
**Tested with Success!**

Verified on device with Android 10
